### PR TITLE
[FIXED JENKINS-34370] Fix NPE getting size of parameterDefinitions.size

### PR DESCRIPTION
**Pull request by fbelzunc**
Fix NPE getting size of parameterDefinitions.size

https://issues.jenkins-ci.org/browse/JENKINS-34370

See https://github.com/jenkinsci/build-pipeline-plugin/pull/98 as well

@reviewbybees CC @carlosrodlop @recena

### CODE REVIEW

**Comment by recena** [comment]
@fbelzunc I don't like the idea using the same issue ID for both PR. In my opinion, this PR requieres a new Jira issue.

**Comment by oleg-nenashev** [comment]
It's just a symptom IMHO. How does this ParameterDefinition manages to get a null list? Do you generate the job somehow from XML?

Also 🐛 , because get() method should be verified as well

**Comment by fbelzunc** [comment]
@oleg-nenashev 

> It's just a symptom IMHO. How does this ParameterDefinition manages to get a null list? Do you generate the job somehow from XML?

Unfortunately, I have no clue why it might happen, but in any case we should about a NullPointerException.

> I don't like the idea using the same issue ID for both PR. In my opinion, this PR requieres a new Jira issue.

I think this is too simple so It doesn't even require a Jenkins Jira issue to be opened.

**Comment by oleg-nenashev** [comment]
Such NPE may be invoked anywhere if you load improper config...

**Comment by recena** [comment]
@fbelzunc 

> I think this is too simple so It doesn't even require a Jenkins Jira issue to be opened.

Up to you. It is only a recommendation that [I received few days ago](https://github.com/jenkinsci/jenkins/pull/2283#issuecomment-213611866) by @daniel-beck.

**Comment by daniel-beck** [comment]
@fbelzunc JIRA issues are a precondition to get fixes backported into LTS (ideally labeled `lts-candidate` so we know). So if that's not a concern, then smaller changes can be accepted without JIRA issue.

**Comment by oleg-nenashev** [comment]
I still keep 🐛 
* If we agree that the null field is a **valid** behavior, it needs to be annotated and documented properly (Javadoc, annotations)
* If we have no clue about the cause, I consider null value as a bug somewhere. In such case let's annotate the field as NonNull and to add the proper list initialization to constructors and readResolve()

**Comment by oleg-nenashev** [comment]
ping @fbelzunc

**Comment by fbelzunc** [comment]
Sorry, I will review this today.

**Comment by fbelzunc** [comment]
@oleg-nenashev Would you mind to review this, please? notice that AFAIK to use readResolve I need to rewrite the variable to use a not final value.

**Comment by oleg-nenashev** [comment]
> notice that AFAIK to use readResolve I need to rewrite the variable to use a not final value.

Yes

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 1]
```diff
@@ -66,15 +66,30 @@
...
+            this.parameterDefinitions = parameterDefinitions;
+        } else {
+            this.parameterDefinitions = Collections.emptyList();
+        }
     }
 
-    public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions) {
-        this.parameterDefinitions = Arrays.asList(parameterDefinitions);
+    public ParametersDefinitionProperty(@Nonnull ParameterDefinition... parameterDefinitions) {
+        if (parameterDefinitions != null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: DRY
use <code>this(parameterDefinitions != null ? Arrays.asList(parameterDefinitions) : null)</code>

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 2]
```diff
@@ -66,15 +66,30 @@
...
         implements Action {
 
-    private final List<ParameterDefinition> parameterDefinitions;
+    private List<ParameterDefinition> parameterDefinitions;
 
     @DataBoundConstructor
-    public ParametersDefinitionProperty(List<ParameterDefinition> parameterDefinitions) {
-        this.parameterDefinitions = parameterDefinitions;
+    public ParametersDefinitionProperty(@Nonnull List<ParameterDefinition> parameterDefinitions) {
+        if (parameterDefinitions != null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Now it's <code>@CheckForNull</code>

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Could be also simplified via a ternary operator

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 3]
```diff
@@ -66,15 +66,30 @@
...
+            this.parameterDefinitions = parameterDefinitions;
+        } else {
+            this.parameterDefinitions = Collections.emptyList();
+        }
     }
 
-    public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions) {
-        this.parameterDefinitions = Arrays.asList(parameterDefinitions);
+    public ParametersDefinitionProperty(@Nonnull ParameterDefinition... parameterDefinitions) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: CheckForNull as well?

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 4]
```diff
@@ -98,11 +113,19 @@ public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions)
     public List<String> getParameterDefinitionNames() {
         return new AbstractList<String>() {
             public String get(int index) {
-                return parameterDefinitions.get(index).getName();
+                if (parameterDefinitions != null && parameterDefinitions.get(index) != null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: This null check should not be required anymore

**Comment by stephenc** [comment]
or you can clone the instance and return a replacement instance. that way the field can stay final. either choice is supported by the `readResolve()` contract... but the clone version only works if you have the `readResolve()` as `private` because otherwise a superclass will be converted into the parent class when invoking...

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 5]
```diff
@@ -98,11 +113,19 @@ public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions)
...
+                    return parameterDefinitions.get(index).getName();
+                } else {
+                    return "";
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: I'm not sure this is a right return value. Why not null?

**Comment by oleg-nenashev** [comment]
@fbelzunc reviewed

**Comment by fbelzunc** [comment]
I think that now it looks much better ;-) 

Thanks for the review.

**Comment by oleg-nenashev** [comment]
🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 6]
```diff
@@ -66,15 +66,22 @@
...
+        this.parameterDefinitions = (parameterDefinitions != null) ? parameterDefinitions : Collections.<ParameterDefinition>emptyList();
     }
 
-    public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions) {
-        this.parameterDefinitions = Arrays.asList(parameterDefinitions);
+    public ParametersDefinitionProperty(@CheckForNull ParameterDefinition... parameterDefinitions) {
+        this(parameterDefinitions != null ? Arrays.asList(parameterDefinitions) : null);
+    }
+
+    protected Object readResolve() {
```
Comment (jglick) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: :bug: should be `private`

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 7]
```diff
@@ -66,15 +66,22 @@
 public class ParametersDefinitionProperty extends OptionalJobProperty<Job<?, ?>>
         implements Action {
 
-    private final List<ParameterDefinition> parameterDefinitions;
+    private List<ParameterDefinition> parameterDefinitions;
 
     @DataBoundConstructor
-    public ParametersDefinitionProperty(List<ParameterDefinition> parameterDefinitions) {
-        this.parameterDefinitions = parameterDefinitions;
+    public ParametersDefinitionProperty(@CheckForNull List<ParameterDefinition> parameterDefinitions) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: :bug: no, make this `@Nonnull` and throw an exception in the constructor if someone makes this mistake.

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 8]
```diff
@@ -98,7 +105,11 @@ public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions)
     public List<String> getParameterDefinitionNames() {
         return new AbstractList<String>() {
             public String get(int index) {
-                return parameterDefinitions.get(index).getName();
+                if (parameterDefinitions.get(index) != null) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: :ant: better to check for null elements in the constructor and `readResolve`

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java (hunk 6 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: And if we are making it private then just `return parameterDefinitions == null ? new ParametersDefinitionProperty() : this;` and the field can stay final FTW

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java (hunk 7 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Agree

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java (hunk 8 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Agree

**Comment by stephenc** [comment]
:bug: as there is a fix that retains the field as final

**Comment by fbelzunc** [comment]
What do you think now?

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 9]
```diff
@@ -69,12 +69,22 @@
     private final List<ParameterDefinition> parameterDefinitions;
 
     @DataBoundConstructor
-    public ParametersDefinitionProperty(List<ParameterDefinition> parameterDefinitions) {
+    public ParametersDefinitionProperty(@Nonnull List<ParameterDefinition> parameterDefinitions) {
+        if (parameterDefinitions == null) {
+            throw new IllegalArgumentException("ParameterDefinitions is null when this is a not valid value");
```
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Throw an a NPE if null not an IAE which is reserved for a non-null value that is not correct

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java [hunk 10]
```diff
@@ -69,12 +69,22 @@
...
+        }
         this.parameterDefinitions = parameterDefinitions;
     }
 
-    public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions) {
-        this.parameterDefinitions = Arrays.asList(parameterDefinitions);
+    public ParametersDefinitionProperty(@Nonnull ParameterDefinition... parameterDefinitions) {
+        if (parameterDefinitions == null) {
+            throw new IllegalArgumentException("ParameterDefinitions is null when this is a not valid value");
```
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: NPE

File: core/src/main/java/hudson/model/ParametersDefinitionProperty.java (hunk 3 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Nope Nonnull please

Comment (fbelzunc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: @stephenc @jglick told me to do it - why it is wrong?

Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: I'm saying this should be `@Nonnull` which is what you have, ignore the CheckForNull zealot... we can argue about which `@Nonnull` or `@NonNull` but that is a different argument and this is Jenkins core which has a habit of using the wrong annotation so we'll not fight that fight here

Comment (stephenc) [inline core/src/main/java/hudson/model/ParametersDefinitionProperty.java]: Though now you have changed this again... what happened to just throwing a `NullPointerException` in the constructor?

**Comment by stephenc** [comment]
:bee:

**Comment by jglick** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
🐝 and 👍 
Byt @fbelzunc should be now ready to NPEs in other stuff if it really misuses it
@reviewbybees done

**Pull request merged**